external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/InternalKafkaConsumerPool.scala

```diff
@@ -0,0 +1,260 @@
+   */
+  def invalidateKey(key: CacheKey): Unit = {
+    // invalidate all idle consumers for the key
+    pool.clear(key)
+
+    pool.getNumActive()
+    // set invalidate timestamp to let active objects being destroyed when returning to pool
+    objectFactory.keyToLastInvalidatedTimestamp.put(key, System.currentTimeMillis())
+  }
+
+  /**
+   * Closes the keyed object pool. Once the pool is closed,
+   * borrowObject will fail with [[IllegalStateException]], but returnObject and invalidateObject
+   * will continue to work, with returned objects destroyed on return.
+   *
+   * Also destroys idle instances in the pool.
+   */
+  def close(): Unit = {
+    pool.close()
+  }
+
+  def getNumIdle: Int = pool.getNumIdle
+
+  def getNumIdle(key: CacheKey): Int = pool.getNumIdle(key)
+
+  def getNumActive: Int = pool.getNumActive
+
+  def getNumActive(key: CacheKey): Int = pool.getNumActive(key)
+
+  def getTotal: Int = getNumIdle + getNumActive
+
+  def getTotal(key: CacheKey): Int = getNumIdle(key) + getNumActive(key)
+
+  private def updateKafkaParamForKey(key: CacheKey, kafkaParams: ju.Map[String, Object]): Unit = {
+    // We can assume that kafkaParam should not be different for same cache key,
+    // otherwise we can't reuse the cached object and cache key should contain kafkaParam.
+    // So it should be safe to put the key/value pair only when the key doesn't exist.
+    objectFactory.keyToKafkaParams.putIfAbsent(key, kafkaParams)
+  }
+
+  private def extractCacheKey(intConsumer: InternalKafkaConsumer): CacheKey = {
+    new CacheKey(intConsumer.topicPartition, intConsumer.kafkaParams)
+  }
+}
+
+private[kafka010] object InternalKafkaConsumerPool {
+
+  /**
+   * Builds the pool for [[InternalKafkaConsumer]]. The pool instance is created per each call.
+   */
+  def build: InternalKafkaConsumerPool = {
+    val objFactory = new ObjectFactory
+    val poolConfig = new PoolConfig
+    new InternalKafkaConsumerPool(objFactory, poolConfig)
+  }
+
+  case class PooledObjectInvalidated(key: CacheKey, lastInvalidatedTimestamp: Long,
+                                     lastBorrowedTime: Long) extends RuntimeException
+
+  object CustomSwallowedExceptionListener extends SwallowedExceptionListener with Logging {
+    override def onSwallowException(e: Exception): Unit = e match {
+      case e1: PooledObjectInvalidated =>
+        logDebug("Pool for key was invalidated after cached object was borrowed. " +
+          s"Invalidating cached object - key: ${e1.key} / borrowed timestamp: " +
+          s"${e1.lastBorrowedTime} / invalidated timestamp for key: ${e1.lastInvalidatedTimestamp}")
+
+      case _ => logError(s"Error closing Kafka consumer", e)
+    }
+  }
+
+  class PoolConfig extends GenericKeyedObjectPoolConfig[InternalKafkaConsumer] {
+    init()
+
+    def init(): Unit = {
+      import PoolConfig._
+
+      val conf = SparkEnv.get.conf
+      val capacity = conf.getInt(CONFIG_NAME_CAPACITY, DEFAULT_VALUE_CAPACITY)
```

Comment:
Hmm... I think `spark.sql.kafkaConsumerCache.capacity` wasn't documented somewhere, but it may be the thing to fix. The patch will provide more configurations to let end users can tune, so maybe worth to add them to some docs? Not 100% sure where to.